mlpipeline/pipeline/graph_utils.py

```diff
@@ -0,0 +1,240 @@
+        """
+
+        graph = self._create_graph(components, inputs, outputs)
+        running_order = self._get_topological_order(graph)
+        return running_order
+
+    def _create_graph(self, components: dict, inputs: set, outputs: set) -> dict:
+        """Creates the dependency graph between components, where node is component id and edge represents the link between them.
+
+        Parameters
+        ----------
+        components : dict
+            The components info
+
+        inputs: set
+            The set of pipeline inputs
+        outputs: set
+            The set of pipeline outputs
+
+        Returns
+        -------
+        dict
+            graph representation of components
+
+        Raises
+        ------
+        RuntimeError
+            If the inputs of one component don"t match with the output of the component it depends on
+            OR inputs and outputs in incorrect format
+        """
+
+        graph = {}
+        for component_id in components:
+            dependencies = components[component_id].dependencies
+            if component_id not in graph:
+                graph[component_id] = set()
+
+            for dependency in dependencies:
+                if dependency not in graph:
+                    graph[dependency] = set()
+
+                if self._verify_edge(components, inputs, outputs, dependency, component_id):
+                    graph[dependency].add(component_id)
+                else:
+                    L.error("Incorrect graph, shutting down")
+                    raise RuntimeError("Incorrect link from {} to {}, inputs don't match the outputs".format(dependency, component_id))
+
+        if len(outputs) > 0:
+            graph["outputs"] = set()
+
+        for output in outputs:
+            output_split = output.split(".")
+            if len(output_split) == 1:
+                if output_split not in inputs:
+                    L.error("Incorrect outputs, shutting down")
+                    raise RuntimeError("Incorrect outputs {}".format(outputs))
+
+                graph["inputs"].add("outputs")
+
+            elif len(output_split) == 2:
+                if self._verify_edge(components, inputs, outputs, output_split[0], "outputs"):
+                    graph[output_split[0]].add("outputs")
+                else:
+                    L.error("Incorrect outputs, shutting down")
+                    raise RuntimeError("Incorrect outputs {}".format(outputs))
+
+            else:
+                L.error("Incorrect inputs, shutting down")
+                raise RuntimeError("Incorrect input {}, should be <component_id>.<input_name>".format(output))
+
+        return graph
+
+    def _get_topological_order(self, graph: dict) -> list:
+        """Gets the dependency graph topologically sorted and creates an order of components to be executed in optimal order.
+        Checks if the graph contains a cycle (if output of one component is referenced somewhere in its" deps)
```

Comment:
Nitpick: it might be beneficial to mention the the algorithm is using DFS to determine the topological order.